### PR TITLE
Manage snippet registry to write snippet in JSON

### LINKED ISSUE #640: Manage snippet registry to write snippet in JSON
Today all snippets (cdata, comment, Insert DTD Entity, etc) are managed with Java code. It should be more clean and will facilitate integration of another snippet (like  https://github.com/eclipse/lemminx/issues/536) if we could write snippets in JSON vscode format.

### PR DESCRIPTION
**Pull request by angelozerr**
Fixes #640

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

**Comment by angelozerr** [comment]
This PR manages snippets with vscode JSON format. In other words

 * existing snippets:

  * CDATA snippets
  * Comments snippets
  * DTD ELEMENT, ENTITY, ATTLIST snippets (inside DTD or DOCTYPE)

 * new snippets for doctype.

Here some advanced snippet support:

 * snippet are displayed according some condition (ex : DOCTYPE snippet is working only if file is empty or if completion is trigger before the root element which doesn't declare a DOCTPE.
 * when DOCTYPE snippet is applied, it uses teh root element name to generate the proper DOCTYPE.

**Comment by angelozerr** [comment]
Here a demo:

DocTypeSnippetDemo

**Comment by angelozerr** [comment]
@fbricon @dakwon you can play now with my PR, I need to clean my code and add some other tests.

**Comment by fbricon** [comment]
xml prolog snippet should be the 1st choice in an empty xml document. It's totally missing

**Comment by fbricon** [comment]
xml prolog should also probably be added to all snippets

**Comment by fbricon** [comment]
If there's an xml prolog already, the doctype snippets are not even proposed on the next line

**Comment by xorye** [comment]
I'm very excited for this PR.

For the prolog snippet named "Insert xml Processing Instruction with standalone", one of the placeholders are appearing:
image

Also, I think the space before the `?>` at the end should be gone?

**Comment by angelozerr** [comment]
> xml prolog should also probably be added to all snippets

done

> If there's an xml prolog already, the doctype snippets are not even proposed on the next line

fixed

>For the prolog snippet named "Insert xml Processing Instruction with standalone", one of the placeholders are appearing:

fixed

> Also, I think the space before the ?> at the end should be gone?

fixed

@xorye could you play again with this PR and give me feedback, thanks!

**Comment by angelozerr** [comment]
> xml prolog snippet should be the 1st choice in an empty xml document. It's totally missing

I have not managed sort for te moment, could we see that in an another PR please.

**Comment by angelozerr** [comment]
Code should be cleaned and test are writen.

**Comment by angelozerr** [comment]
Here a new demo:

SnippetDemo

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/commons/snippets/SnippetRegistry.java [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+	 * Returns the snippet completion items according to the context filter.
+	 * 
+	 * @param replaceRange       the replace range.
+	 * @param lineDelimiter      the line delimiter.
+	 * @param canSupportMarkdown true if markdown is supported to generate
+	 *                           documentation and false otherwise.
+	 * @param contextFilter      the context filter.
+	 * @return the snippet completion items according to the context filter.
+	 */
+	public List<CompletionItem> getCompletionItems(Range replaceRange, String lineDelimiter, boolean canSupportMarkdown,
```
Comment (xorye) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/commons/snippets/SnippetRegistry.java:233]: I think after #680 and #669 are both merged, we should use `xml.preferences.quoteStyle` to determine whether or not to use `"` or `'` in the insert snippet text. I can create an issue for that.

**Comment by xorye** [comment]
The features in the demo gif work for me.

I have some feedback:

When I get completion after typing `<?`, do we still need the `<?xml ... ?>` option?
image

Also, when I select `Insert xml Processing Instruction` or `Insert xml Processing Instruction with standalone`, I think there is a problem with the replace range:
image
There is an extra `<?`. I only get this problem with the prolog snippets.

Also if completion is invoked inside a comment like this:
image

I don't think we should be getting any snippets in this case

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/snippets/PrologSnippetContext.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+			return false;
+		}
+		Position start = request.getReplaceRange().getStart();
+		if (start.getLine() > 0) {
+			// The xml processing instruction must be declared in the first line.
+			return false;
+		}
+		// No xml processing instruction, check if completion was triggered before the
+		// document element
+		DOMElement documentElement = document.getDocumentElement();
```
Comment (xorye) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/snippets/PrologSnippetContext.java:45]: Can we also check for DocType? If I invoke completion at `<!DOCTYPE |root-element [`, I don't think I should get the prolog snippets.

image

**Comment by fbricon** [comment]
Please fix the merge conflict

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/commons/snippets/SnippetRegistry.java (hunk 1 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/commons/snippets/SnippetRegistry.java:233]: Yes please create an issue.

File: org.eclipse.lemminx/src/main/resources/org/eclipse/lemminx/services/snippets/doctype-snippets.json [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  "Insert  DOCTYPE with subset": {
+    "prefix": [
+      "<!DOCTYPE"
+    ],
+    "body": [
+      "<!DOCTYPE ${1:root-element} [",
+      "\t<!ELEMENT ${1:root-element} (${2:#PCDATA}) >",
+      "]>"
+    ],
+    "description": "Insert  DOCTYPE with subset"
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/resources/org/eclipse/lemminx/services/snippets/doctype-snippets.json]: @xorye fixed

**Comment by angelozerr** [comment]
> xml prolog should also probably be added to all snippets

It can be complicated to do that because we need to manage case of existing or not of xml processing instruction to avoid generating twice.

I had the problem with DOCTYPE (which generates a root-element if doens't exist and don't generate it otherwise). To do that I have duplicated snippets (see newfile-snippets.json and doctype-snippets.json).

To manage properly this usecase we should manage aggregation of several snippets, but perhaps we could do that in the future? 

> Also, when I select Insert xml Processing Instruction or Insert xml Processing Instruction with standalone, I think there is a problem with the replace range:
There is an extra <?. I only get this problem with the prolog snippets.

@xorye fixed

> Also if completion is invoked inside a comment like this:
I don't think we should be getting any snippets in this case

@xorye fixed

> Can we also check for DocType? If I invoke completion at <!DOCTYPE |root-element [, I don't think I should get the prolog snippets.

@xorye fixed

> Please fix the merge conflict

@fbricon  fixed

The JSON snippet needs to be improved again, but if you are OK could we done that in separate issues:

 * gives the capability to sort the snippets (xml prolog snippet should be the 1st choice in an empty xml document. )
 * eat the '>' when it exists. I mean when we have `<? | >` and xml processing instruction is applied, it doesn't eat the `>` and it generates

```xml
<?xml version="1.0" encoding="UTF-8"?>>
```
 * snippet should take care of quote preferences.

**Comment by xorye** [comment]
Great, the fixes work on my machine.

I found a small inconsistency, XML should either always be upper case or always be lower case. 
image
image

File: org.eclipse.lemminx/src/main/resources/org/eclipse/lemminx/services/snippets/doctype-snippets.json [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    "description": "Insert PUBLIC DOCTYPE with subset"
+  },
+  "Insert DOCTYPE with subset": {
+    "prefix": [
+      "<!DOCTYPE"
+    ],
+    "suffix": ">",
+    "body": [
+      "<!DOCTYPE ${1:root-element} [",
+      "\t<!ELEMENT ${1:root-element} (${2:#PCDATA}) >",
```
Comment (xorye) [inline org.eclipse.lemminx/src/main/resources/org/eclipse/lemminx/services/snippets/doctype-snippets.json:53]: Remove extra space before the closing `>`

File: org.eclipse.lemminx/src/main/resources/org/eclipse/lemminx/services/snippets/doctype-snippets.json [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    ],
+    "description": "Insert SYSTEM DOCTYPE"
+  },
+  "Insert PUBLIC DOCTYPE": {
+    "prefix": [
+      "<!DOCTYPE"
+    ],
+    "suffix": ">",
+    "body": [
+      "<!DOCTYPE ${1:root-element} PUBLIC \"${2:public-id}\" \"${3:file}.dtd\" >"
```
Comment (xorye) [inline org.eclipse.lemminx/src/main/resources/org/eclipse/lemminx/services/snippets/doctype-snippets.json:18]: Remove extra space before the closing `>`

File: org.eclipse.lemminx/src/main/resources/org/eclipse/lemminx/services/snippets/new-xml-snippets.json [hunk 6]
```diff
@@ -0,0 +1,97 @@
+{
+  "New XML with SYSTEM DOCTYPE": {
+    "prefix": [
+      "<!DOCTYPE"
+    ],
+    "suffix": ">",
+    "body": [
+      "<!DOCTYPE ${1:root-element} SYSTEM \"${2:file}.dtd\" >",
```
Comment (xorye) [inline org.eclipse.lemminx/src/main/resources/org/eclipse/lemminx/services/snippets/new-xml-snippets.json:8]: Remove extra space before the closing `>`

File: org.eclipse.lemminx/src/main/resources/org/eclipse/lemminx/services/snippets/new-xml-snippets.json [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    ],
+    "description": "New XML with SYSTEM DOCTYPE"
+  },
+  "New XML with PUBLIC DOCTYPE": {
+    "prefix": [
+      "<!DOCTYPE"
+    ],
+    "suffix": ">",
+    "body": [
+      "<!DOCTYPE ${1:root-element} PUBLIC \"${2:public-id}\" \"${3:file}.dtd\" >",
```
Comment (xorye) [inline org.eclipse.lemminx/src/main/resources/org/eclipse/lemminx/services/snippets/new-xml-snippets.json:20]: Remove extra space before the closing `>`

File: org.eclipse.lemminx/src/main/resources/org/eclipse/lemminx/services/snippets/new-xml-snippets.json [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+    "description": "New XML with PUBLIC DOCTYPE with subset"
+  },
+  "New XML with DOCTYPE with subset": {
+    "prefix": [
+      "<!DOCTYPE"
+    ],
+    "suffix": ">",
+    "body": [
+      "<!DOCTYPE ${1:root-element} [",
+      "\t<!ELEMENT ${1:root-element} (${2:#PCDATA}) >",
```
Comment (xorye) [inline org.eclipse.lemminx/src/main/resources/org/eclipse/lemminx/services/snippets/new-xml-snippets.json:61]: Remove extra space before the closing `>`

**Comment by xorye** [comment]
Also, the completion options are a bit hard to identify IMO:
image

I think the old descriptions were nicer:
image

**Comment by xorye** [comment]
For the noNamespaceSchemaLocation snippet, the attributes are on a separate line:
```
<root-element
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:noNamespaceSchemaLocation="file.xsd">
    
</root-element>
```

A cool enhancement (maybe this can be done on a different issue) would be to have the attributes on a separate line if `xml.format.splitAttributes` is true. 

Taking it one step further, it might be cool to have snippet text be preformatted against the formatter. 
Because if the user has format on save enabled, and the user saves the document after inserting a snippet, I think it would be weird if the snippet content's formatting changes dramatically (like in the case for split attributes)

File: org.eclipse.lemminx/src/main/resources/org/eclipse/lemminx/services/snippets/doctype-snippets.json (hunk 4 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/resources/org/eclipse/lemminx/services/snippets/doctype-snippets.json]: fixed

File: org.eclipse.lemminx/src/main/resources/org/eclipse/lemminx/services/snippets/doctype-snippets.json (hunk 5 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/resources/org/eclipse/lemminx/services/snippets/doctype-snippets.json]: fixed

File: org.eclipse.lemminx/src/main/resources/org/eclipse/lemminx/services/snippets/new-xml-snippets.json (hunk 6 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/resources/org/eclipse/lemminx/services/snippets/new-xml-snippets.json]: fixed

File: org.eclipse.lemminx/src/main/resources/org/eclipse/lemminx/services/snippets/new-xml-snippets.json (hunk 7 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/resources/org/eclipse/lemminx/services/snippets/new-xml-snippets.json]: fixed

File: org.eclipse.lemminx/src/main/resources/org/eclipse/lemminx/services/snippets/new-xml-snippets.json (hunk 8 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/resources/org/eclipse/lemminx/services/snippets/new-xml-snippets.json]: fixed

**Comment by angelozerr** [comment]
> I found a small inconsistency, XML should either always be upper case or always be lower case.

fixed

**Comment by angelozerr** [comment]
> But things like this could be discussed in another issue.

I agree with you we should pay attention with format settings in otherwise snippet should be formatted (each time or better one time and cache the formatted snippet).

Please create an issue for that. Thanks

**Comment by fbricon** [comment]
Jenkins fails https://ci.eclipse.org/lemminx/blue/organizations/jenkins/lemminx/detail/PR-680/34/pipeline:
```
[ERROR] Failures: 

[ERROR]   XMLCompletionSnippetsTest.afterComment:118 expected: <<!DOCTYPE root-element SYSTEM "file.dtd" >

<root-element>

</root-element>> but was: <<!DOCTYPE root-element SYSTEM "file.dtd">

<root-element>

</root-element>>

[ERROR]   XMLCompletionSnippetsTest.afterProlog:133 expected: <<!DOCTYPE root-element SYSTEM "file.dtd" >

<root-element>

</root-element>> but was: <<!DOCTYPE root-element SYSTEM "file.dtd">

<root-element>

</root-element>>

[ERROR]   XMLCompletionSnippetsTest.emptyXMLContent:38 expected: <<!DOCTYPE root-element SYSTEM "file.dtd" >

<root-element>

</root-element>> but was: <<!DOCTYPE root-element SYSTEM "file.dtd">

<root-element>

</root-element>>
```

**Comment by angelozerr** [comment]
@fbricon fixed

**Comment by xorye** [comment]
The snippet is eating more than the `>`:

gif

If this is not a big problem, I'm good with this PR being merged

**Comment by angelozerr** [comment]
> If this is not a big problem, I'm good with this PR being merged

No it's very bad -( Let me try to fix it.

**Pull request merged**